test/e2e/app/app.go

```diff
@@ -321,6 +376,55 @@ func (app *Application) ProcessProposal(req abci.RequestProcessProposal) abci.Re
 	return abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_ACCEPT}
 }
 
+// ExtendVote will produce vote extensions in the form of random numbers to
+// demonstrate vote extension nondeterminism.
+//
+// In the next block, if there are any vote extensions from the previous block,
+// a new transaction will be proposed that updates a special value in the
+// key/value store ("extensionSum") with the sum of all of the numbers collected
+// from the vote extensions.
+func (app *Application) ExtendVote(req abci.RequestExtendVote) abci.ResponseExtendVote {
+	// We ignore any requests for vote extensions that don't match our expected
+	// next height.
+	if req.Height != int64(app.state.Height)+1 {
+		return abci.ResponseExtendVote{}
```

Comment:
This shouldn't happen (?). Consider adding a log message

Reply:
Should there be any consequences for the application here? i.e. should it be standard practice to panic here, or just ignore these extensions? I suppose we could make it application-specific and let the application developers choose what to do with it, but under what circumstances would Tendermint be sending invalid `ExtendVote` requests to the application?

Reply:
As discussed live, either there is a bug in Tendermint, or there is a bug in the app. So up to the app to decide what to do here. So up to you :-)

Reply:
If it ignores the height mismatch error, would it be possible to see it in any of the logs (for debugging propose when something is wrong)?

Reply:
We're now logging this error in the E2E testing app.